src/annotator/test/frame-observer-test.js

```diff
@@ -4,128 +4,230 @@ import {
-    const frame = createAnnotatableIFrame();
-    await waitForFrameObserver();
-
-    assert.calledOnce(onFrameAdded);
-    assert.calledWith(onFrameAdded, frame);
-
-    frame.setAttribute('src', document.location);
-    await waitForIFrameUnload(frame);
-
-    assert.calledOnce(onFrameRemoved);
-    assert.calledWith(onFrameRemoved, frame);
-  });
-
-  it(`doesn't call onFrameAdded if FrameObserver is disconnected`, async () => {
-    frameObserver.disconnect();
-    const frame = createAnnotatableIFrame();
-
-    frameObserver._discoverFrames(); // Emulate a race condition
-    await onDocumentReady(frame);
-
-    assert.notCalled(onFrameAdded);
+  describe('onDocumentReady', () => {
+    let clock;
+    let fakeIFrame;
+    let fakeIFrameDocument;
+
+    class FakeIFrameDocument extends EventTarget {
+      constructor() {
+        super();
+        this.readyState = 'loading';
+        this.location = {
+          href: 'about:blank',
+        };
+      }
+    }
+
+    beforeEach(() => {
+      clock = sinon.useFakeTimers();
+      fakeIFrameDocument = new FakeIFrameDocument();
+      fakeIFrame = document.createElement('div');
+      fakeIFrame.contentWindow = { document: fakeIFrameDocument };
+      fakeIFrame.setAttribute('src', 'http://my.dummy');
+    });
+
+    afterEach(() => {
+      clock.restore();
+    });
+
+    it('waits for the iframe to change the location and readyState if the initial document is blank', () => {
+      fakeIFrameDocument.location.href = 'about:blank';
+      const onReady = onDocumentReady(fakeIFrame);
+
+      // After the initial 'about:blank' document, a new document is loaded.
+      const newDocument = new FakeIFrameDocument();
+      newDocument.location.href = 'http://my.dummy';
+      newDocument.readyState = 'complete';
+      fakeIFrame.contentWindow = { document: newDocument };
+      clock.tick(10);
+
+      return onReady;
+    });
+
+    it('rejects promise if the document stays in the initial blank document too long', () => {
+      fakeIFrameDocument.location.href = 'about:blank';
+      const promise = onDocumentReady(fakeIFrame).catch(error =>
+        assert.equal(error.message, 'Annotatable iframe took too long to load')
+      );
+
+      clock.tick(1000);
+
+      return promise;
+    });
+
+    it('waits for the iframe DOMContentLoaded event to be triggered if the document is loading', () => {
+      fakeIFrameDocument.location.href = 'about:srcdoc';
+      fakeIFrameDocument.readyState = 'loading';
+      const onDOMContentLoaded = onDocumentReady(fakeIFrame);
+
+      fakeIFrameDocument.dispatchEvent(new Event('DOMContentLoaded'));
+
+      return onDOMContentLoaded;
+    });
+
+    it("resolves immediately if document is 'complete' or 'interactive'", () => {
+      fakeIFrameDocument.location.href = 'about:srcdoc';
+      fakeIFrameDocument.readyState = 'complete';
+      const onReady = onDocumentReady(fakeIFrame);
```

Comment:
```suggestion
      return onDocumentReady(fakeIFrame);
```

`onReady` variable seems superfluous here.